src/main/java/org/fcrepo/storage/ocfl/DefaultOcflObjectSession.java

```diff
@@ -0,0 +1,477 @@
+        return readStreamOptional(path)
+                .orElseThrow(() -> new NotFoundException(String.format("File %s was not found for resource %s",
+                        path, resourceId)));
+    }
+
+    private Optional<InputStream> readStreamOptional(final PathPair path) {
+        if (isOpen() && (deleteObject || deletePaths.contains(path))) {
+            return Optional.empty();
+        }
+
+        return readFromStaging(path).or(() -> readFromOcfl(path));
+    }
+
+    private Optional<InputStream> readFromStaging(final PathPair path) {
+        final var stagingPath = stagingPath(path);
+
+        if (Files.exists(stagingPath)) {
+            try {
+                return Optional.of(Files.newInputStream(stagingPath));
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    private Optional<InputStream> readFromOcfl(final PathPair path) {
+        if (ocflRepo.containsObject(ocflObjectId)) {
+            final var object = ocflRepo.getObject(ObjectVersionId.head(ocflObjectId));
+            if (object.containsFile(path.path)) {
+                return Optional.of(object.getFile(path.path).getStream());
+            }
+        }
+        return Optional.empty();
+    }
+
+    private Path stagingPath(final PathPair path) {
+        return objectStaging.resolve(path.encodedPath);
+    }
+
+    private Path createStagingPath(final PathPair path) {
+        final var stagingPath = stagingPath(path);
+
+        try {
+            Files.createDirectories(stagingPath.getParent());
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+
+        return stagingPath;
+    }
+
+    private void write(final Optional<InputStream> content, final Path destination) {
+        if (content.isPresent()) {
+            try {
+                Files.copy(content.get(), destination, StandardCopyOption.REPLACE_EXISTING);
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }
+    }
+
+    private void writeHeaders(final ResourceHeaders headers, final Path destination) {
+        try {
+            objectMapper.writeValue(destination.toFile(), headers);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private void deletePathsFromStaging() {
+        deletePaths.stream().map(this::stagingPath).forEach(path -> {
+            if (Files.exists(path)) {
+                try {
+                    Files.delete(path);
+                } catch (IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+            }
+        });
+    }
+
+    private ResourceHeaders readHeaders(final InputStream stream) {
+        try {
+            return objectMapper.readValue(stream, ResourceHeaders.class);
```

Comment:
Minor, but I believe we can optimize this some by initializing `ObjectReader` and `ObjectWriter`s for `ResourceHeaders` per session rather than making new ones for each read/write.

`ObjectReader headerReader = objectMapper.readerFor(ResourceHeaders.class)`
etc.

Reply:
done